src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs

```diff
@@ -1148,22 +1147,28 @@ internal SingleLookupResult CheckViability(Symbol symbol, int arity, LookupOptio
                 return LookupResult.Inaccessible(symbol, diagInfo);
             }
             else if (!InCref &&
-                     !this.IsAccessible(unwrappedSymbol,
+                    !this.IsAccessible(unwrappedSymbol,
                                         RefineAccessThroughType(options, accessThroughType),
                                         out inaccessibleViaQualifier,
                                         ref useSiteDiagnostics,
                                         basesBeingResolved))
             {
-                if (inaccessibleViaQualifier)
-                {
-                    diagInfo = diagnose ? new CSDiagnosticInfo(ErrorCode.ERR_BadProtectedAccess, unwrappedSymbol, accessThroughType, this.ContainingType) : null;
-                }
-                else
-                {
-                    var unwrappedSymbols = ImmutableArray.Create<Symbol>(unwrappedSymbol);
-                    diagInfo = diagnose ? new CSDiagnosticInfo(ErrorCode.ERR_BadAccess, new[] { unwrappedSymbol }, unwrappedSymbols, additionalLocations: ImmutableArray<Location>.Empty) : null;
-                }
-
+                 bool friendRefNotEqualToThis = unwrappedSymbol != null && unwrappedSymbol.ContainingAssembly.GetInternalsVisibleToPublicKeys(this.ContainingType.ContainingAssembly.Name).Any()
+                            && unwrappedSymbol.DeclaredAccessibility.Equals(Accessibility.Internal);
+
+                 foreach (ImmutableArray<byte> key in unwrappedSymbol.ContainingAssembly.GetInternalsVisibleToPublicKeys(this.ContainingType.ContainingAssembly.Name))
+                 {
+                     friendRefNotEqualToThis = key.SequenceEqual(this.ContainingType.ContainingAssembly.Identity.PublicKey) ? false : friendRefNotEqualToThis;
+                 }
+
+                 diagInfo = diagnose ?
```

Comment:
I'm not sure this is what @cston had in mind. I think he just wanted the `diagnose` check lifted to before the first `if` statement, but I may be wrong. complete conversion to `? :` doesn't look more readable to me.